libs/algo/odc/algo/_percentile.py

```diff
@@ -0,0 +1,58 @@
+from typing import List, Sequence, Tuple
+import dask.array as da
+import xarray as xr
+import numpy as np
+from ._geomedian import reshape_for_geomedian
+from ._masking import keep_good_np
+
+
+def np_percentile(xx, percentile, nodata):
+
+    mask = xx != nodata
+    valid_counts = mask.sum(axis=-1)
+
+    sorted_arr = np.sort(xx, axis=-1)
+    index_offsets = np.arange(0, xx.size, xx.shape[-1])
+
+    indices = np.round(percentile * (valid_counts - 1)).astype(np.int64).flatten()
+    indices = indices + index_offsets
+
+    xx = sorted_arr.take(indices).reshape(xx.shape[:-1])
+
+    return keep_good_np(xx, (valid_counts >= 3), nodata)
+
+
+def xr_percentile(
+    src: xr.Dataset,
+    percentiles: Sequence,
+    nodata,
+) -> xr.Dataset:
+
+    """
+    Calculates the percentiles of the input data along the time dimension.
+
+    :param src: xr.Dataset, bands can be either
+        float or integer with `nodata` values to indicate gaps in data.
+        `nodata` must be the largest values in the dataset or NaN.
+
+    :param compute_mads: A sequence of percentiles in the [0.0, 1.0] range
+
+    :param nodata: The `nodata` value
+    """
+
+    xx = reshape_for_geomedian(src)
```

Comment:
I don't think this a good idea. This reshape is quite expensive to perform on large data and it introduces dependencies between bands, which is not needed, you can compute percentiles on every band separately, right?

I say we should keep the same axis order and process every band separately, this thing should operate on an Array not on a Dataset. It would be nice to still accept Dataset in Dataset out for convenience, but per band operations should be independent Dask graphs.

Expected input axis order is T,Y,X single band, output is Y,X,P or Y,X if only one percentile was requested.

So really you only need to make sure that T dimension is one single chunk, by rechunking if necessary.

Reply:
I've timed both this and my original and interestingly they take about the same amount of time. I think the cache friendliness of my approach (~10x `np_percedntile`) compensates for the `rechunk...` cost. 

But your suggestion has a way nicer dask graph and the code is simpler too 👍

Reply:
Dask is hard to benchmark, performance changes a lot depending on size and number of worker and everything. I can see how transpose can help with sorting performance, but there is absolutely no need to interleave bands and make them data dependent without reason.